### PR TITLE
fix: fail unsafe keras h5 lambda ambiguity

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary

- Restrict passed H5 Lambda code-analysis checks to explicit safe normalization allowlist patterns.
- Treat syntax-valid but non-allowlisted Lambda code as a failed warning instead of “safe Python code.”
- Add a computed builtins-dispatch regression that previously bypassed direct dangerous-construct matching.

## Validation

- `PROMPTFOO_DISABLE_TELEMETRY=1 PYTHONPATH=/private/tmp/modelaudit-c031 /Users/mdangelo/code/modelaudit/.venv/bin/pytest tests/scanners/test_keras_h5_scanner.py::test_lambda_dynamic_builtins_dispatch_is_not_marked_safe tests/scanners/test_keras_h5_scanner.py::test_lambda_safe_normalization_pattern_still_passes tests/scanners/test_keras_h5_scanner.py::test_lambda_additional_safe_normalization_patterns_still_pass tests/scanners/test_keras_h5_scanner.py::test_lambda_safe_prefix_with_injected_code_is_flagged -q`
- `PROMPTFOO_DISABLE_TELEMETRY=1 PYTHONPATH=/private/tmp/modelaudit-c031 /Users/mdangelo/code/modelaudit/.venv/bin/pytest tests/scanners/test_keras_h5_scanner.py -q`
- `PROMPTFOO_DISABLE_TELEMETRY=1 PYTHONPATH=/private/tmp/modelaudit-c031 /Users/mdangelo/code/modelaudit/.venv/bin/ruff format --check modelaudit/scanners/keras_h5_scanner.py tests/scanners/test_keras_h5_scanner.py`
- `PROMPTFOO_DISABLE_TELEMETRY=1 PYTHONPATH=/private/tmp/modelaudit-c031 /Users/mdangelo/code/modelaudit/.venv/bin/ruff check modelaudit/scanners/keras_h5_scanner.py tests/scanners/test_keras_h5_scanner.py`
- `PROMPTFOO_DISABLE_TELEMETRY=1 PYTHONPATH=/private/tmp/modelaudit-c031 /Users/mdangelo/code/modelaudit/.venv/bin/mypy modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/`
- `git diff --check`

## Scan Finding

- Fixes MA-DSS-C031 from the Codex Security scan.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
@codex review

File: modelaudit/scanners/keras_h5_scanner.py [hunk 1]
```diff
@@ -1075,38 +1130,56 @@ def _check_lambda_layer(self, layer_config: dict[str, Any], result: ScanResult)
...
+                        details={
+                            "layer_class": "Lambda",
+                            "validation_status": "invalid_python",
+                            "validation_error": error,
+                        },
+                        why="Malformed Lambda source cannot be safely classified.",
+                        rule_code="S507",
+                    )
+                    function_requires_review = True
+        elif function_str is not None:
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Fixed in fe6e563e. The malformed scalar branch now explicitly excludes dict/list values already dispatched to the recognized serialized-function handlers. Added regressions for both recognized dict and list bytecode to ensure neither emits the duplicate malformed-metadata finding. Local H5 tests, Ruff, mypy, and the full non-slow suite pass through test execution; awaiting exact-head CI before resolving.

**Comment by mldangelo-oai** [comment]
QA pass on exact head `c16bb6905c8435b5e2a2d32a908ace88eb6f3a22`.

Additional fixes pushed while reviewing:
- Tightened Lambda source allowlisting so arbitrary `K.<name>(x)` / `tf.nn.<name>(x)` helpers no longer pass as `safe_normalization`; added regressions for `K.get_value(x)` and `tf.nn.embedding_lookup(x)` while keeping `tf.nn.softmax(x)` / `K.softmax(x)` green.
- Hardened Lambda module/function reference detection for dotted dangerous function names; `compat.v1.py_func` and `io.system` now report CRITICAL, while near-match `io.systematic_helper` stays WARNING via token-boundary coverage.

Validation:
- `pytest tests/scanners/test_keras_h5_scanner.py -q` -> 125 passed
- `ruff check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/` -> clean
- `ruff format --check modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/` -> clean
- `mypy modelaudit/ packages/modelaudit-picklescan/src packages/modelaudit-picklescan/tests tests/` -> clean
- `pytest -n auto -m "not slow and not integration" --maxfail=1` -> 8090 passed, 17 skipped
- GitHub checks are green on `c16bb6905c8435b5e2a2d32a908ace88eb6f3a22`.

**Comment by mldangelo-oai** [comment]
QA follow-up on exact head `1f8098c9`.

Additional fixes pushed during adversarial review:
- Replaced the regex-only source allowlist with AST shape validation, so malformed Unicode/newline lookalikes cannot receive a passing safety check.
- Accepted benign finite float normalization while rejecting zero-divisor expressions.
- Replaced the module-root/function cross-product with explicit module/function pairs; trusted-looking descendants and invalid pairings now fail closed.
- Routed legacy `function_type="function"` H5 metadata through the module reference allowlist, avoiding false positives for canonical named callables such as `keras.activations.relu`.

Focused validation:
- `PROMPTFOO_DISABLE_TELEMETRY=1 .venv/bin/pytest tests/scanners/test_keras_h5_scanner.py -q` -> 136 passed
- `.venv/bin/ruff check modelaudit/scanners/keras_h5_scanner.py tests/scanners/test_keras_h5_scanner.py` -> clean
- `.venv/bin/mypy modelaudit/scanners/keras_h5_scanner.py tests/scanners/test_keras_h5_scanner.py` -> clean

All existing review threads are resolved. CI is running on the new head.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 2]
```diff
@@ -1036,9 +1112,9 @@ def _check_lambda_layer(self, layer_config: dict[str, Any], result: ScanResult)
             )
 
         # Check if there's actual Python code to validate
-        if function_str and isinstance(function_str, str):
-            # First check if it matches safe patterns
-            is_safe_pattern = any(pattern.match(function_str.strip()) for pattern in self._SAFE_LAMBDA_PATTERNS)
+        if function_str and isinstance(function_str, str) and not is_named_function_reference:
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Fixed in 40385b83. Named-reference handling is now limited to the canonical legacy shape where the function field itself is the reference and no separate function_name exists. Conflicting metadata validates the function string as source and suppresses an allowlisted pass when review is required. Added a malicious-source regression.

[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Fixed in 40385b83. String values are now validated regardless of truthiness, so an empty function field fails closed as malformed Python source. Added a regression that asserts the failed invalid_python check.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 3]
```diff
@@ -82,8 +127,14 @@ class KerasH5Scanner(BaseScanner):
...
             "eval",
             "exec",
+            "load_file_system_library",
+            "load_library",
+            "load_op_library",
             "open",
             "popen",
+            "numpy_function",
+            "py_func",
+            "py_function",
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py:202]: Fixed in 40385b83. Added the normalized raw-op tokens pyfunc, pyfuncstateless, and eagerpyfunc to the dangerous callback set. Added regressions for all three aliases plus a benign PyFunctional near-match to guard against overmatching.

**Comment by mldangelo-oai** [comment]
Addressed all three open review findings in 40385b83:\n\n- conflicting Lambda metadata can no longer suppress source analysis\n- empty function strings now fail closed\n- TensorFlow raw PyFunc aliases are CRITICAL, with a benign near-match regression\n\nValidation: 142 Keras H5 scanner tests passed; Ruff and mypy passed on the touched files.

**Comment by mldangelo-oai** [comment]
QA update on `b6d9d88e` after merging current `main`:

- Preserved the fail-closed Lambda source/module policy and all prior review fixes.
- Removed two concrete false positives: signed finite normalization literals (while signed zero divisors remain rejected), and exact legacy TensorFlow Keras callable modules emitted by the official serializer (`tensorflow.python.keras.activations` / `tensorflow.python.keras.backend`).
- Added exact-module near-match coverage so descendants such as `tensorflow.python.keras.activations.attacker` remain findings.
- Associated validation only: `tests/scanners/test_keras_h5_scanner.py` -> 167 passed. Scoped Ruff, format check, mypy, and `git diff --check` are clean.

I did not run the full suite locally; CI is handling the complete matrix.

**Comment by mldangelo-oai** [comment]
Fresh critical review on current `main` found and fixed three adjacent issues in `66711bfa`:\n\n- empty/whitespace legacy `function_type="function"` values no longer bypass both source and module-reference findings\n- safe AST allowlist checks now bind to the Lambda’s actual single parameter name, removing false positives such as `lambda inputs: inputs / 255.0` without widening the accepted expression shapes\n- exact CPython native execution modules now receive critical treatment, while dotted descendants, case/prefix lookalikes, and benign `operator` references remain warnings rather than false criticals\n\nValidation with the declared `h5` extra:\n- full changed Keras H5 scanner module: 194 passed\n- scoped Ruff format/check: clean\n- scoped mypy: clean\n- `git diff --check`: clean\n\nNo unresolved review threads remain. Moving on while fresh CI runs.

**Comment by mldangelo-oai** [comment]
Pushed `5900ac63` after adversarial review.

Fixes:
- addresses all three open review threads
- inspects executable legacy `output_shape`/`mask` callbacks and Keras 3 dict-format `output_shape`
- preserves raw shape and modern raw-mask behavior to avoid false positives
- requires canonical safe module/function names
- restricts qualified Lambda recognition to known framework paths

Focused validation:
- 28 targeted H5 Lambda regressions passed
- scoped Ruff lint/format passed
- scoped mypy passed

CI is running on the new head; moving to the next PR in parallel.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 4]
```diff
@@ -66,30 +68,106 @@ class KerasH5Scanner(BaseScanner):
...
+        "tensorflow": frozenset({"abs", "identity", "sigmoid", "tanh"}),
+        "tensorflow.math": frozenset({"abs", "sigmoid", "tanh"}),
+        "tensorflow.nn": _SAFE_TF_NN_LAMBDA_FUNCTIONS,
+        "tensorflow.python.keras.activations": frozenset(
+            {"elu", "gelu", "hard_sigmoid", "relu", "selu", "sigmoid", "softmax", "softplus", "softsign", "tanh"}
+        ),
+        "tensorflow.python.keras.backend": _SAFE_K_BACKEND_LAMBDA_FUNCTIONS,
+        "tf_keras.backend": _SAFE_K_BACKEND_LAMBDA_FUNCTIONS,
+    }
+    _QUALIFIED_LAMBDA_LAYER_CLASSES: ClassVar[frozenset[str]] = frozenset(
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Fixed in 74958cb5. Lambda class recognition now accepts any dotted Lambda class under exact trusted Keras framework roots while rejecting custom namespace lookalikes. Added trusted-submodule positives, custom-prefix negatives, and an H5 dangerous-bytecode integration regression.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 5]
```diff
@@ -1155,27 +1308,59 @@ def _check_lambda_layer(self, layer_config: dict[str, Any], result: ScanResult)
...
+                    why=get_pattern_explanation("lambda_layer"),
+                    rule_code="S1103",
+                )
+        if callback_field == "function":
+            for auxiliary_field in ("output_shape", "mask"):
+                auxiliary_value = layer_config.get(auxiliary_field)
+                is_serialized_lambda = (
+                    auxiliary_field == "output_shape"
+                    and isinstance(auxiliary_value, dict)
+                    and auxiliary_value.get("class_name") == "__lambda__"
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Fixed in 74958cb5. Dict-format serialized Lambda callbacks are now recognized for both output_shape and mask without requiring legacy type markers. Added h5py-independent critical detection coverage and an H5 integration regression.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 6]
```diff
@@ -1024,39 +1111,86 @@ def _scan_wrapped_layer_config(self, layer_config: Any, result: ScanResult) -> N
...
         if isinstance(function_str, dict):
             encoded_function_handled = check_lambda_dict_function(
                 function_str,
                 result,
                 self.current_file_path,
-                layer_name,
+                callback_layer_name,
             )
+            if not encoded_function_handled:
+                function_requires_review = True
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Fixed in 74958cb5. Unrecognized function dictionaries now extract nested module and function references from function_name or string config fields before the generic fallback. Dangerous references are critical, allowlisted framework references pass, substring near-matches stay warning-level, and a dangerous nested reference suppresses misleading safe sibling passes.

File: modelaudit/scanners/keras_h5_scanner.py [hunk 7]
```diff
@@ -1184,22 +1386,19 @@ def _is_lambda_layer_class(layer_class: Any) -> bool:
         if not isinstance(layer_class, str):
             return False
 
-        normalized = layer_class.strip()
-        if normalized == "Lambda":
+        if layer_class == "Lambda":
             return True
+        return layer_class.endswith(".Lambda") and layer_class.startswith(KerasH5Scanner._TRUSTED_LAMBDA_LAYER_PREFIXES)
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/keras_h5_scanner.py]: Addressed in 6609b0ec. Lambda CVE attribution now uses an explicit set of known serialized Keras/TensorFlow layer class paths instead of trusting every keras.* prefix. Added attacker-submodule and lookalike negatives while retaining historical experimental and legacy positives; 18 focused helper tests passed with Ruff and mypy clean.
